### PR TITLE
MNT: Run PTH flake test in prep for supporting pathlib (utils 1/n) (⏰  wait for #16930)

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description
Extracted from #16060
~`utils` is a big one so I'm focusing on a single module for this PR in order to unblock #16927~

blocked by 
- #16930
- #16931 

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: astropy/utils/data.py [hunk 1]
```diff
@@ -967,7 +965,7 @@ def compute_hash(localfn):
 
     Parameters
     ----------
-    localfn : str
+    localfn : str | Path
```
Comment (pllim) [inline astropy/utils/data.py:968]: This looks like a new feature? I think this one actually needs a change log.

Comment (neutrinoceros) [reply astropy/utils/data.py:968]: good point. I need to review this thoroughly myself before I undraft, and I'll keep this in mind when I do !

File: astropy/utils/data.py [hunk 2]
```diff
@@ -339,6 +330,10 @@ def get_readable_fileobj(
     if isinstance(name_or_obj, os.PathLike):
         name_or_obj = os.fspath(name_or_obj)
 
+    # FIXME: this should be fixed
```
Comment (pllim) [inline astropy/utils/data.py:333]: Can you please elaborate on what exactly should be fixed? Is there a follow-up issue?

File: astropy/utils/data.py [hunk 3]
```diff
@@ -651,7 +646,7 @@ def get_pkg_data_filename(
 
     Parameters
     ----------
-    data_name : str
+    data_name : str | Path
```
Comment (pllim) [inline astropy/utils/data.py:649]: This is considered new feature, so probably needs a change log. Same comment applies to all other similar occurrences in this PR (i.e., adding Path support to public API).

File: astropy/utils/data.py [hunk 4]
```diff
@@ -2180,13 +2180,20 @@ def cache_contents(pkgname="astropy"):
...
     except OSError:
         return _NOTHING
-    with os.scandir(dldir) as it:
-        for entry in it:
-            if entry.is_dir:
-                url = get_file_contents(
-                    os.path.join(dldir, entry.name, "url"), encoding="utf-8"
-                )
-                r[url] = os.path.abspath(os.path.join(dldir, entry.name, "contents"))
+    # with os.scandir(dldir) as it:
```
Comment (pllim) [inline astropy/utils/data.py:2183]: Why is this commented?

File: astropy/utils/tests/test_data.py [hunk 5]
```diff
@@ -1473,22 +1467,21 @@ def test_check_download_cache_cleanup(temp_cache, valid_urls):
...
-    with open(bf3, "w") as f:
-        f.write("awkwardly-named bogus file that exists")
+    bf3 = dldir / "contents"
+    bf3.write_text("awkwardly-named bogus file that exists")
 
     u2, c2 = next(valid_urls)
     f2 = download_file(u, cache=True)
-    os.unlink(f2)
-    bf4 = os.path.dirname(f2)
+    # os.unlink(f2)
```
Comment (pllim) [inline astropy/utils/tests/test_data.py:1481]: Why is this commented?

**Review by pllim — COMMENTED** [review]
I am a little worried about changing so much low level utility code. Is ruff really that hard to please?

This one might need feedback from downstream like `astroquery` (@bsipocz) and HPC users (@aarchiba , @weaverba137 , etc).

**Comment by weaverba137** [comment]
I'm not aware of any serious HPC issues with Path objects, but I haven't researched the topic extensively.

**Comment by neutrinoceros** [comment]
This PR isn't ready for review yet.
@pllim I'm having a lot of the same questions you raised, I intend to answer them before I undraft this, but I'll prioritize fixing CI first.

**Comment by pllim** [comment]
There are conflicts now and looks like one of the blockers still unresolved, so moving milestone.

**Comment by neutrinoceros** [comment]
while resolving merge conflicts I realized this is full of unwarranted API changes so I think I'll just drop this PR for now.

**Pull request closed without merging**